commons/src/main/java/org/infinispan/commons/util/concurrent/BlockingAbortPolicy.java

```diff
@@ -0,0 +1,27 @@
+package org.infinispan.commons.util.concurrent;
+
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * A handler for rejected tasks that runs the task if the current thread is a blocking thread otherwise it
+ * rejects the task.
+ * @author wburns
+ * @since 10.1
+ */
+public class BlockingAbortPolicy extends ThreadPoolExecutor.AbortPolicy {
+   private BlockingAbortPolicy() { }
+
+   private static final BlockingAbortPolicy INSTANCE = new BlockingAbortPolicy();
+
+   public static ThreadPoolExecutor.AbortPolicy getInstance() {
+      return INSTANCE;
+   }
+
+   @Override
+   public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+      if (executor.isShutdown() || Thread.currentThread().getName().startsWith("async")) {
```

Comment:
Also a trace log would be nice.

The `startsWith` check looks a bit like a hack, maybe we should use a custom `Thread` subclass? Maybe not in this PR though :)

Reply:
Sure, I can add a trace.

Yes, I definitely want to add a custom thread subclass. Just gotta get everything ready for this :)